### DISCUSSION TITLE
onFailure hooks?

### DISCUSSION THREAD

**Discussion by agardnerIT** [devcontainers/community#115] [Q&A]
Is there a way to hook into the failure event for a devcontainer? I'd like to send a cloudevent if a container creation fails.

**Comment by samruddhikhandale** [comment]
Currently, Dev Container spec does not provide a built-in way to hook into the failure event of a container creation. However, you can achieve this by using Docker event monitoring and a script to send a cloudevent when a container fails to start.

Here is a high-level approach:

1. Monitor Docker events: You can use the docker events command to monitor events on the Docker daemon. You can filter these events to only get events related to container creation.
2. Detect failure: When a container fails to start, Docker generates an event with the status die or health_status: unhealthy. You can detect these events and trigger your script.
3. Send cloudevent: Once a failure is detected, you can use a script to send a cloudevent. This script can be written in any language that you prefer and can make HTTP requests.

Here is a simple example using a bash script:

```
#!/bin/bash

docker events --filter 'type=container' --filter 'event=die' --filter 'event=health_status: unhealthy' |
while read -r line
do
    CONTAINER_ID=$(echo $line | awk '{print $2}')
    CONTAINER_NAME=$(docker inspect --format '{{.Name}}' $CONTAINER_ID)
    # Replace with your code to send a cloudevent
    echo "Container $CONTAINER_NAME has failed"
done

```

This script will monitor Docker events and print a message when a container fails to start. You can replace the echo command with your code to send a cloudevent.

Please note that this is a workaround and might not cover all scenarios. For a more robust solution, you might need to use a Docker monitoring tool or service.
